### PR TITLE
libusermode: refactor log strategies

### PR DESCRIPTION
**Pull request by chivay**
In current implementation, log strategy is implemented as a std::string.
Using an enum is both faster and less error-prone.

Additionaly, remove the concept of "stack" strategy as it has no sense
(doing work, without any benefit to the user).

### CODE REVIEW

File: src/libusermode/userhook.cpp [hunk 1]
```diff
@@ -916,20 +931,25 @@ void drakvuf_load_dll_hook_config(drakvuf_t drakvuf, const char* dll_hooks_list_
...
             e.type = HOOK_BY_OFFSET;
 
-            if (!std::getline(ss, e.log_strategy, ',') || e.log_strategy.empty())
+            std::string strategy_name;
+            if (!std::getline(ss, strategy_name, ',') || strategy_name.empty())
                 throw -1;
 
-            if (e.log_strategy != "log" && e.log_strategy != "log+stack" && e.log_strategy != "stack")
+            strategy = get_log_strategy(strategy_name);
+            if (!strategy.has_value())
```
Comment (BonusPlay) [inline src/libusermode/userhook.cpp:949]:
```suggestion
            if (!strategy)
```

**Review by BonusPlay — APPROVED** [review]
LGTM

**Comment by icedevml** [comment]
@drakvuf-jenkins This is OK to test.

File: src/libusermode/userhook.cpp [hunk 2]
```diff
@@ -855,6 +856,20 @@ bool drakvuf_request_usermode_hook(drakvuf_t drakvuf, const dll_view_t* dll, con
     return true;
 }
 
+std::optional<LogStrategy> get_log_strategy(const std::string& str)
```
Comment (kscieslinski) [inline src/libusermode/userhook.cpp:859]: I don't think LogStrategy is a good name. My understanding of `dll_hooks` is that sometimes plugins authors want to leave some space for the end-users to define which calls should trigger which actions (example `log`, `stack`). And so in `apimon` we want the user to be able to customize the list of apicalls which should be logged and in `memdump` we want the user to be able to customize the list of apicalls which will perform the `stack walk`.

Comment (chivay) [reply src/libusermode/userhook.cpp:859]: I'm open to suggestions and personally not really satisfied with the name, but I wasn't able to come up with anything better ;)

Comment (disaykin) [reply src/libusermode/userhook.cpp:859]: Maybe "trace"?

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

File: src/libusermode/userhook.cpp (hunk 2 above)
Comment (chivay) [reply src/libusermode/userhook.cpp]: After a discussion with @kscieslinski, I've decided to make it a `struct` to be more easily extendable in the future.

**Review by kscieslinski — APPROVED** [review]
I really like how it looks now

**Review by icedevml — APPROVED** [review]
ack, looks better now

**Pull request merged**